corehq/apps/enterprise/iterators.py

```diff
@@ -1,7 +1,242 @@
+            'domain': domain
+        }
+
+    @classmethod
+    def get_query_paraams(cls, fetched_object):
+        '''
+        Takes a fetched, converted object and returns the values from this object that will be necessary
+        to continue where this query left off.
+        '''
+        return {
+            'domain': fetched_object['domain'],
+            'inserted_at': fetched_object['inserted_at'],
+            'id': fetched_object['form_id']
+        }
+
+    @classmethod
+    def get_kwargs_from_map(cls, map):
+        '''
+        Takes a map-like object from a continuation request (generally GET/POST) and extracts
+        the parameters necessary for initializing an IterableEnterpriseFormQuery.
+        '''
+        last_domain = map.get('domain', None)
+        last_time = map.get('inserted_at', None)
+        if last_time:
+            last_time = datetime.fromisoformat(last_time).astimezone(timezone.utc)
+        last_id = map.get('id', None)
+        return {
+            'last_domain': last_domain,
+            'last_time': last_time,
+            'last_id': last_id
+        }
+
+
+class AppIdToNameResolver:
+    def __init__(self):
+        self.domain_lookup_tables = {}
+
+    def resolve_app_id_to_name(self, domain, app_id):
+        if 'domain' not in self.domain_lookup_tables:
+            domain_apps = get_brief_apps_in_domain(domain)
+            self.domain_lookup_tables[domain] = {a.id: a.name for a in domain_apps}
+
+        return self.domain_lookup_tables[domain].get(app_id, None)
+
+
+def run_query_over_domains(query_factory, domains, limit=None, last_domain=None, **kwargs):
+    iterators = []
+    if last_domain:
+        remaining_domains = dropwhile(lambda d: d != last_domain, domains)
+        in_progress_domain = next(remaining_domains)
+        iterators.append(run_query_over_domain(query_factory, in_progress_domain, limit=limit, **kwargs))
+    else:
+        remaining_domains = domains
+
+    fresh_domain_query_args = query_factory.get_next_query_args(kwargs, last_hit=None)
+    for domain in remaining_domains:
+        iterators.append(run_query_over_domain(query_factory, domain, limit=limit, **fresh_domain_query_args))
+
+    yield from islice(chain.from_iterable(iterators), limit)
+
+
+def run_query_over_domain(query_factory, domain, limit=None, **kwargs):
+    remaining = limit
+
+    next_query_args = kwargs
+
+    while True:
+        query = query_factory.get_query(domain, limit=limit, **next_query_args)
+        results = query.run()
+        for hit in results.hits:
+            last_hit = hit
+            yield last_hit
+
+        num_fetched = len(results.hits)
+
+        if num_fetched >= results.total or (remaining and num_fetched >= remaining):
+            break
+        else:
+            if remaining:
+                remaining -= num_fetched
+
+            next_query_args = query_factory.get_next_query_args(next_query_args, last_hit)
+
+
+class ReportQueryFactoryInterface:
```

Comment:
maybe good to use `ABC` here?

Reply:
Addressed in e55664e614d470c8f231c580edd2a6a45b49d180